### PR TITLE
EKF: reconcile lane switching logic between copter and EKF code

### PR DESCRIPTION
**Pull request by tridge**
This prevents an EKF failsafe in copter when we have another good EKF lane

### CODE REVIEW

**Comment by rmackay9** [comment]
Discussed on the dev call and I think we're keen to get @priseborough to give this his OK before merging..

**Comment by priseborough** [comment]
I will go through the code, but we do need to test this. It should be possible to trigger an EKF failsafe only on the first lane in SITL by adding some test code that rotates the magnetomer data for that lane.

**Comment by tridge** [comment]
@priseborough I tested it in SITL by changing the accel offsets on the first IMU while flying. I adjusted the offset so the EKF didn't go unhealthy, but copter EKF failsafe threshold was reached. It did switch lanes correctly to the good IMU when the counter reached 9

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp [hunk 1]
```diff
@@ -789,12 +789,52 @@ void NavEKF3::UpdateFilter(void)
...
+    }
+
+    // update the yaw and position reset data to capture changes due to the lane switch
+    if (newPrimaryIndex != primary) {
+        updateLaneSwitchYawResetData(newPrimaryIndex, primary);
+        updateLaneSwitchPosResetData(newPrimaryIndex, primary);
+        updateLaneSwitchPosDownResetData(newPrimaryIndex, primary);
+        primary = newPrimaryIndex;
+        lastLaneSwitch_ms = now;
+        gcs().send_text(MAV_SEVERITY_CRITICAL, "NavEKF2: lane switch %u", primary);
```
Comment (priseborough) [inline libraries/AP_NavEKF3/AP_NavEKF3.cpp]: Copy and paste error - should be NavEKF3

**Comment by OXINARF** [comment]
Shouldn't this be part of the existing EKF lane switching? What is happening here is "vehicle is checking EKF state and asking it to change lane if possible", which looks like something the EKF should be able to 
do on its own. Or are you saying this specific lane switching is only valid for Copter?

**Comment by priseborough** [comment]
The EKF  lacks context to know whether its  performance level is adequate and the built in criteria is fairly relaxed and not suitable for all situations.

**Comment by priseborough** [comment]
I think the checkLaneSwitch() function being something that we can use in other platform types as required.

**Comment by priseborough** [comment]
Improvements to internal checking are still required however.

File: libraries/AP_NavEKF3/AP_NavEKF3.cpp (hunk 1 above)
Comment (tridge) [reply libraries/AP_NavEKF3/AP_NavEKF3.cpp]: thanks, fixed

**Pull request merged**